sys/net/application_layer/nanocoap/nanocoap.c

```diff
@@ -595,32 +605,106 @@ size_t coap_put_option_uri(uint8_t *buf, uint16_t lastonum, const char *uri, uin
     return bufpos - buf;
 }
 
+size_t coap_block2_init(uint8_t* buf, uint16_t lastonum, coap_pkt_t *pkt, coap_blockhelper_t *blk)
+{
+    uint32_t blknum;
+    unsigned szx;
+    if(coap_get_blockopt(pkt, COAP_OPT_BLOCK2, &blknum, &szx) >= 0) {
+        /* Use the smallest block size */
+        if (NANOCOAP_BLOCK_SZX_MAX - 4 < szx ) {
+            szx =NANOCOAP_BLOCK_SZX_MAX - 4;
+        }
+    }
+    blk->opt = buf;
+    blk->start = blknum * coap_szx2size(szx);
+    blk->end = blk->start + coap_szx2size(szx);
+    blk->cur = 0;
+    return coap_put_option_block2(buf, lastonum, blknum, szx, 1);
+}
+
+ssize_t coap_block2_build_reply(coap_pkt_t *pkt, unsigned code,
```

Comment:
Why not refactor the code here to create a coap_block2_finish() function. This would pair semantically with coap_block2_init(), and allow gcoap to use the function as well. Then this function, if still worthwhile, could just call coap_block2_finish. Please see comment about the blockwise put functions though.

Reply:
I think I did what you requested. Please let me know if the new situation is usable for gcoap

Reply:
Exactly what I was thinking, thanks.